dashboard/lib/services/child_account/grace_period_handler.rb

```diff
@@ -0,0 +1,45 @@
+module Services
+  module ChildAccount
+    class GracePeriodHandler < Services::Base
+      LockoutSchedulingError = Class.new(StandardError)
+
+      attr_reader :user
+
+      # @param user [User] the student account
+      def initialize(user:)
+        @user = user
+      end
+
+      def call
+        return false unless eligible?
+
+        user.transaction do
+          Services::ChildAccount.start_grace_period(user)
+
+          estimated_lockout_date = Policies::ChildAccount.lockout_date(user)
+          raise LockoutSchedulingError, 'Lockout date is not set' unless estimated_lockout_date
+
+          CAP::LockoutJob.set(wait_until: estimated_lockout_date).perform_later(user_id: user.id)
+        end
+
+        true
+      end
+
+      # Checks if the user can be transited to the grace period state.
+      private def eligible?
+        return false if Policies::ChildAccount.compliant?(user)
+        return false if Policies::ChildAccount::ComplianceState.locked_out?(user)
+        # If the user is already in the grace period, it cannot be started again.
+        return false if Policies::ChildAccount::ComplianceState.grace_period?(user)
```

Comment:
Rather then checking all the states they shouldn't be, would it be simpler to instead be explicit about the state they can be?
```suggestion
        return false unless [nil, 's'].include?(user.child_account_compliance_state)
```
A user can only transition from `nil` -> `grace_period` or `request_sent` -> `grace_period`

Reply:
It's a bit more complicated: https://github.com/code-dot-org/code-dot-org/pull/59241/commits/cfd1806231b86bd4cb302f143818ea3eed806a46

Reply:
I read through the edge case you pointed out. I still wonder if the code would be easier to read if it was explicit about the start and end states. For example (but you don't need to copy this exactly):
```
if (start_state_1)
  transition(new_state) if (should_transition_from_state_1)
elsif(start_state_2)
  transition(new_state) if (should_transition_from_state_2)
...
```
Ultimately, this is feedback about readability, not functionality. If you think the code would be harder to read, then go ahead and leave it how you have it. I just wanted to offer a different perspective on how it could have been written.

Reply:
I'm not sure I understand this correctly, but doesn't this add even more conditions that need to be updated each time a new state is added?
I think the only thing that could really help is moving everything to be managed by a state machine.